pkg/storegateway/bucket_index_postings.go

```diff
@@ -37,10 +38,15 @@ func newPostingGroup(addAll bool, addKeys, removeKeys []labels.Label) *postingGr
 }
 
 // NOTE: Derived from tsdb.postingsForMatcher. index.Merge is equivalent to map duplication.
-func toPostingGroup(lvr labelValuesReader, m *labels.Matcher) (*postingGroup, error) {
+func toPostingGroup(lvr indexheader.Reader, m *labels.Matcher) (*postingGroup, error) {
 	if setMatches := m.SetMatches(); len(setMatches) > 0 && (m.Type == labels.MatchRegexp || m.Type == labels.MatchNotRegexp) {
 		keys := make([]labels.Label, 0, len(setMatches))
 		for _, val := range setMatches {
+			if _, err := lvr.PostingsOffset(m.Name, val); errors.Is(err, indexheader.NotFoundRangeErr) {
```

Comment:
WDYT about defining a helper function to make the code cleaner?

```go
func labelValueHasPostings(lvr indexreader.Reader, name, value string) bool {
        _, err := lvr.PostingsOffset(m.Name, val)
        return errors.Is(err, indexheader.NotFoundRangeErr)
}
```

And then:

```suggestion
			if !labelValueHasPostings(lvr, m.Name, val) {
```